### PR TITLE
Make the Leaf dynamics differentiable

### PR DESCRIPTION
**Pull request by SarahAlidoost**
relates #8

### CODE REVIEW

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+        # create a model and optimizer
+        model = DiffLeafDynamics(params, wdp, agro, config_path, external_states)
+        tdwi = torch.nn.Parameter(torch.tensor(0.2, dtype=torch.float32))
+        output = model({"TDWI": tdwi})
+        lai = output[0, :, 0]
+        loss = lai.sum()
+        grads = torch.autograd.grad(loss, tdwi, retain_graph=True)[0]  # this is ∂loss/∂tdwi
+
+        assert grads is not None, "Gradients for TDWI should not be None"
```
Comment (cwmeijer) [inline tests/physical_models/crop/test_leaf_dynamics.py:148]: Maybe a bit overkill for now but I recently learned about subtasks. https://pypi.org/project/pytest-subtests/
This would the code to continue running even after failing a single assert so all checks will be run always.
Something to try later maybe.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        DALV = torch.tensor(0., dtype=DTYPE)
+        if p.SPAN.requires_grad:  # replacing hard threshold `if lvage > p.SPAN``
+            sharpness = 1000.0  # FIXEME
+            for lv, lvage in zip(s.LV, s.LVAGE, strict=False):
+                weight = torch.sigmoid((lvage - p.SPAN) * sharpness)
+                DALV = DALV + weight * lv
+        else:
+            for lv, lvage in zip(s.LV, s.LVAGE, strict=False):
+                if lvage > p.SPAN:
+                    DALV = DALV + lv
```
Comment (cwmeijer) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py:248]: I'm not sure about this approach. The code's behavior is now dependent on whether or not grad is required for SPAN. That isn't very transparent, but it could fit the current situation of the project. Mut maybe we should discuss this.
An alternative would be to have 2 separate implementations for either case. This would cause code duplication of course which we could solve as well, but that could get a little convoluted as well.

**Review by cwmeijer — APPROVED** [review]
I think this is a great PR! 🎖Very neat!
I made some comments. Some are suggestions for possible improvements and others are more points for you to think about (or discuss together) before you decide to merge.
I hope they are helpful!

File: tests/physical_models/crop/test_leaf_dynamics.py (hunk 1 above)
Comment (SarahAlidoost) [reply tests/physical_models/crop/test_leaf_dynamics.py:211]: nice suggestion! for now I skip adding this library as a dependency. I will add it as an issue to look at it later.

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    agro_management_inputs = inputs["AgroManagement"]
+    cropd = inputs["ModelParameters"]
+
+    weather_data_provider = WeatherDataProviderTestHelper(inputs["WeatherVariables"])
+    crop_model_params_provider = ParameterProvider(cropdata=cropd)
+    external_states = inputs["ExternalStates"]
+
+    # convert parameters to tensors
+    crop_model_params_provider.clear_override()
+    for name in ["SPAN", "TDWI", "TBASE", "PERDL", "RGRLAI"]:
```
Comment (cwmeijer) [inline tests/physical_models/crop/test_leaf_dynamics.py:28]: Why is this list hardcoded? Can't you iterate over the keys in the provider? In that case 2 extra parameters would be iterated over: SLATB and KDIFTB

Comment (SarahAlidoost) [reply tests/physical_models/crop/test_leaf_dynamics.py:28]: The types of the parameters `SLATB` and `KDIFTB` are `AfgenTrait` and not `float`, see [here](https://github.com/WUR-AI/diffWOFOST/blob/0984aa5f44c40a1e37528ef4f2916233c303364e/src/diffwofost/physical_models/crop/leaf_dynamics.py#L121-L122). This type is used as some sort of lookup table with interpolation, see class definition [here](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/util.py#L604). Since this PR is about two parameters of `"SPAN" ` and `"TDWI"`, only those that effect the gradient path of `"SPAN" ` and `"TDWI"` are converted to tensor in this PR. All the other parameters should be tested in the next iteration.

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+        assert grad_backward is not None, "Backward gradients for SPAN should not be None"
+        assert grad_backward == grads, "Forward and backward gradients for SPAN should match"
+
+    def test_gradients_span_twlv_leaf_dynamics_numerical(self):
+        # first check if the numerical gradient isnot zero i.e. the parameter has an effect
+        span = torch.nn.Parameter(torch.tensor(30, dtype=torch.float64))
+        numerical_grad = calculate_numerical_grad("SPAN", span, "TWLV")  # this is Δloss/Δspan
+
+        assert numerical_grad == 0.0
```
Comment (michielkallenberg) [inline tests/physical_models/crop/test_leaf_dynamics.py:337]: this test seems not entirely complete

File: tests/physical_models/crop/test_leaf_dynamics.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        output = model({"TDWI": tdwi})
+        twlv = output[0, :, 1]
+        loss = twlv.sum()
+        # this is ∂loss/∂tdwi, for comparison with numerical gradient
+        grads = torch.autograd.grad(loss, tdwi, retain_graph=True)[0]
+
+        assert_almost_equal(numerical_grad, grads.item(), decimal=3)
+
+
+class TestDiffLeafDynamicsSPAN:
```
Comment (michielkallenberg) [inline tests/physical_models/crop/test_leaf_dynamics.py:275]: Conssider making this more generic; there is a bit of code repetion.

File: tests/physical_models/pcse_test_code.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+logging.disable(logging.CRITICAL)
+
+this_dir = os.path.dirname(__file__)
+
+
+def nothing(*args, **kwargs):
+    pass
+
+
+class SimulationObjectTestHelper(SimulationObject):
```
Comment (michielkallenberg) [inline tests/physical_models/pcse_test_code.py:41]: This class seems unused

File: tests/physical_models/pcse_test_code.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        self.AGROMANAGEMENT = AgroManager
+        self.OUTPUT_INTERVAL = "daily"
+        self.OUTPUT_INTERVAL_DAYS = 1
+        self.OUTPUT_WEEKDAY = 0
+        self.OUTPUT_VARS = list(YAML_test_inputs["Precision"].keys())
+        self.SUMMARY_OUTPUT_VARS = []
+        self.TERMINAL_OUTPUT_VARS = []
+
+
+class EngineTestHelper(Engine):
```
Comment (michielkallenberg) [inline tests/physical_models/pcse_test_code.py:148]: There's a bit of code repetition from the original class.

**Review by michielkallenberg — APPROVED** [review]
Very nice work! Great. I left some minor comments; feel very much free to ignore.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    DVS      Crop development stage         DVS_Phenology                    -
+    FL       Fraction biomass to leaves     DVS_Partitioning                 -
+    FR       Fraction biomass to roots      DVS_Partitioning                 -
+    SAI      Stem area index                WOFOST_Stem_Dynamics             -
+    PAI      Pod area index                 WOFOST_Storage_Organ_Dynamics    -
+    TRA      Transpiration rate             Evapotranspiration              |cm day-1| ?
+    TRAMX    Maximum transpiration rate     Evapotranspiration              |cm day-1| ?
+    ADMI     Above-ground dry matter        CropSimulation                  |kg ha-1 d-1|
+             increase
+    RFTRA    Reduction factor for transpiration (wat & ox)      Y             -
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py:107]: Fixed!

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+    def initialize(self, day, kiosk, parvalues):
+        """Initialize the WOFOST_Leaf_Dynamics simulation object.
+
+        :param day: start date of the simulation
+        :param kiosk: variable kiosk of this PCSE  instance
+        :param parvalues: `ParameterProvider` object providing parameters as
+                key/value pairs
+        """
+        self.kiosk = kiosk
+        # TODO check if parvalues are already torch.nn.Parameters
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py:159]: yes, currently, there are no checks on the parameters since we are not using Traits anymore. This should be fixed in the next iterations.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+        states.DWLV = states.DWLV + rates.DRLV
+        states.TWLV = states.WLV + states.DWLV
+
+        # Store final leaf biomass deques
+        self.states.LV = tLV
+        self.states.SLA = tSLA
+        self.states.LVAGE = tLVAGE
+
+    @prepare_states
+    def _set_variable_LAI(self, nLAI):  # FIXEME
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py:348]: This function, as a BMI function, might not be used in any application that requires differentiability. For now, I kept it, but will most probably be removed.

File: tests/physical_models/crop/test_leaf_dynamics.py (hunk 4 above)
Comment (SarahAlidoost) [reply tests/physical_models/crop/test_leaf_dynamics.py:337]: right, it is fixed now.

File: tests/physical_models/pcse_test_code.py (hunk 7 above)
Comment (SarahAlidoost) [reply tests/physical_models/pcse_test_code.py:148]: That's correct. This file will be improved later.
